auth/auth_test.go

```diff
@@ -0,0 +1,257 @@
+		{ConfigPath: cp[6], id: "logger-123", Pass: false},
+		{ConfigPath: cp[7], id: "", Pass: false}, //idk
+	}
+	for tc, _ := range testCases {
+		config, err1 := config.NewConfig(testCases[tc].ConfigPath)
+		if err1 != nil {
+			PrintResult(t, testCases[tc].Pass, err1)
+		} else {
+			au, err2 := NewAuth(config)
+			if err2 != nil {
+				PrintResult(t, testCases[tc].Pass, err2)
+			} else {
+				err3 := au.Auth(testCases[tc].id, methodCases[tc], &objectCases[tc])
+				if err3 != nil {
+					PrintResult(t, testCases[tc].Pass, err3)
+				} else {
+					PrintResult(t, testCases[tc].Pass, err2)
+				}
+			}
+		}
+	}
+}
+
+type TestInput1 struct {
+	tip    TestInput
+	id     string
+	method object.Method
+	key    object.Key
+	Pass   bool
+}
+
+func TestHash(t *testing.T) {
+	testCases := []TestInput1{
+		{tip: TestInput{ConfigPath: cp[0]}, id: "agent-123", method: methodCases[0], key: object.Key{Type: "world", Name: "judge", Namespace: "default"}, Pass: true},
+		{tip: TestInput{ConfigPath: cp[1]}, id: "judge-123", method: methodCases[2], key: object.Key{Type: "action"}, Pass: true},
+		{tip: TestInput{ConfigPath: cp[2]}, id: "logger-123", method: methodCases[1], key: object.Key{}, Pass: true},
+		{tip: TestInput{ConfigPath: cp[6]}, id: "agent-123", method: methodCases[4], key: object.Key{Type: "action"}, Pass: false},
+	}
+	for _, test := range testCases {
+		config, err1 := config.NewConfig(test.tip.ConfigPath)
+		if err1 != nil {
+			PrintResult(t, test.Pass, err1)
+		} else {
+			au, err2 := NewAuth(config)
+			if err2 != nil {
+				PrintResult(t, test.Pass, err2)
+			} else {
+				hash0 := fmt.Sprintf("%s-%s-%s-%s-%s", test.id, test.method, test.key.Type, test.key.Namespace, test.key.Name)
+				hash1 := au.hash(test.id, test.method, test.key)
+				if hash0 != hash1 {
+					fmt.Println(hash0, hash1, "sdf")
+					PrintResult(t, test.Pass, errors.New("The hash is not working properly"))
+				} else {
+					PrintResult(t, test.Pass, nil)
+				}
+			}
+		}
+	}
+}
+
+var objectKeys = []object.Key{
+	//config0
+	object.Key{Type: "world", Name: "wrongName"},
+	object.Key{Type: "action", Namespace: "default"},
+	object.Key{Type: "action"},
+	object.Key{Type: "world"},
+	object.Key{},
+	//config1
+	object.Key{Type: "action", Namespace: "nondefault"},
+	object.Key{Type: "world"},
+	object.Key{Name: "wrongName"},
+	//config2
+	object.Key{Type: "nonaction", Name: "wrongName", Namespace: "nondefault"},
+	object.Key{Type: "", Name: "wrongName", Namespace: "nondefault"},
+	object.Key{Type: "action"},
+	object.Key{Type: "nonaction", Name: "", Namespace: "nondefault"},
+	//config3
+	object.Key{Type: "nonaction", Namespace: "default"},
+	object.Key{},
+	object.Key{Type: "world"},
+
+	object.Key{Type: "nonaction", Name: "wrongName", Namespace: "default"},
+	object.Key{Type: "nonaction"},
+	object.Key{Name: "wrongName", Namespace: "default"},
+}
+var objectCases = []object.Object{
+	//object.Object{},     expected to fail but the program exits (?)
```

Comment:
that would be better if you delete this line

Reply:
Shouldn't the program print a log on this case and continue? It just releases an error and terminates.

Reply:
It is a bug in the `auth` package. We should check the value of `obj.Key` and then use it